src/shared/sleep-config.c

```diff
@@ -143,6 +146,136 @@ int read_battery_capacity_percentage(void) {
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/manufacturer: %m");
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/model_name", &bat_model_name);
+        if (r > 0)
+               sha256_process_bytes(bat_model_name, strlen(bat_model_name), &hash);
+        else if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/model_name is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/model_name: %m");
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/serial_number", &bat_serial_number);
+        if (r > 0)
+               sha256_process_bytes(bat_serial_number, strlen(bat_serial_number), &hash);
+        else if (r == -ENOENT)
+               log_debug_errno(r, "/sys/class/power_supply/BAT0/serial_number is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/serial_number: %m");
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r > 0)
+               sha256_process_bytes(&machine_id, sizeof(machine_id), &hash);
+        else if (r == -ENOENT)
+               log_debug_errno(r, "machine ID is unavailable: %m");
+        else if (r < 0)
+               return log_debug_errno(r, "Failed to get machine ID: %m");
+
+        r = id128_get_product(&product_id);
+        if (r > 0)
+               sha256_process_bytes(&product_id, sizeof(product_id), &hash);
+        else if (r == -ENOENT)
+               log_debug_errno(r, "product_id does not exist: %m");
+        else if (r < 0)
+                return log_debug_errno(r, "Failed to get product ID: %m");
+
+        sha256_finish_ctx(&hash, ret);
+
+        return 0;
+}
+
+/* battery percentage discharge rate per hour is in range 0-200 then return success */
+static int battery_discharge_rate_in_range(int battery_discharge_rate) {
+        if (battery_discharge_rate <= 0 || battery_discharge_rate >= 200)
+               return 0;
+               /* battery discharge rate is not in a valid range */
+        log_debug("Battery discharge rate is in valid range");
+        return 1;
+}
+
+/* Battery percentage discharge rate per hour is read from specific file. It is stored along with system
+ * and battery identifier hash to maintain the integrity of discharge rate value */
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *hash_id_discharge_rate = NULL;
+        const char *p, *filepath;
+        uint8_t current_hash_id[SHA256_DIGEST_SIZE];
+        char *stored_hash_id = NULL, *stored_discharge_rate = NULL, *h = NULL;
+        int discharge_rate, r;
+
+        filepath = "/var/lib/systemd/sleep/battery_discharge_percentage_rate_per_hour";
+
+        r = read_one_line_file(filepath, &hash_id_discharge_rate);
+        log_debug("Stored hash_id and discharge_rate is %s", hash_id_discharge_rate);
```

Comment:
I do not think this debugging log is not necessary...
At least, this must be moved to after the error handling, as `hash_id_discharge_rate` may be NULL.

Reply:
Yes. Another one of my debug logs. Will remove in final push after all reviews.